mitmproxy crashed error after application connects to a websocket

### Discussed in https://github.com/mitmproxy/mitmproxy/discussions/6601

<div type='discussions-op-text'>

<sup>Originally posted by **Botytec** January 14, 2024</sup>
im running mitmproxy in local mode that is proxying only one application, everytime that application connects to a websocket server i get this error

mitmproxy has crashed!
Traceback (most recent call last):
  File "mitmproxy\proxy\server.py", line 380, in server_event
  File "mitmproxy\proxy\layer.py", line 156, in handle_event
  File "mitmproxy\proxy\layer.py", line 269, in handle_event
  File "mitmproxy\proxy\layer.py", line 156, in handle_event
  File "mitmproxy\proxy\tunnel.py", line 98, in _handle_event
  File "mitmproxy\proxy\layers\tls.py", line 485, in event_to_child
  File "mitmproxy\proxy\tunnel.py", line 153, in event_to_child
  File "mitmproxy\proxy\layer.py", line 156, in handle_event
  File "mitmproxy\proxy\tunnel.py", line 98, in _handle_event
  File "mitmproxy\proxy\tunnel.py", line 153, in event_to_child
  File "mitmproxy\proxy\layer.py", line 156, in handle_event
  File "mitmproxy\proxy\layers\http\__init__.py", line 880, in _handle_event
  File "mitmproxy\proxy\layers\http\__init__.py", line 936, in event_to_child
  File "mitmproxy\proxy\layer.py", line 147, in handle_event
  File "mitmproxy\proxy\utils.py", line 26, in _check_event_type
AssertionError: Unexpected event type at HttpStream._handle_event: Expected Start|HttpEvent, got Reply(WebsocketStartHook(flow=<HTTPFlow
</div>

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -9,6 +9,7 @@
 
 * Fix a regression where clientplayback would break due to eager task execution.
   ([#6605](https://github.com/mitmproxy/mitmproxy/pull/6605), @mhils)
+* Fix a regression where WebSocket connections would break due to eager task execution.
 * Fix bug where insecure HTTP requests are saved incorrectly when exporting to HAR files.
   ([#6578](https://github.com/mitmproxy/mitmproxy/pull/6578), @DaniElectra)
 

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -10,6 +10,7 @@
 * Fix a regression where clientplayback would break due to eager task execution.
   ([#6605](https://github.com/mitmproxy/mitmproxy/pull/6605), @mhils)
 * Fix a regression where WebSocket connections would break due to eager task execution.
+  ([#6609](https://github.com/mitmproxy/mitmproxy/pull/6609), @mhils)
 * Fix bug where insecure HTTP requests are saved incorrectly when exporting to HAR files.
   ([#6578](https://github.com/mitmproxy/mitmproxy/pull/6578), @DaniElectra)
 

--- a/mitmproxy/proxy/layers/http/__init__.py
+++ b/mitmproxy/proxy/layers/http/__init__.py
@@ -528,8 +528,8 @@ class HttpStream(layer.Layer):
                 yield commands.Log(
                     f"{self.debug}[http] upgrading to {self.child_layer}", DEBUG
                 )
-            yield from self.child_layer.handle_event(events.Start())
             self._handle_event = self.passthrough
+            yield from self.child_layer.handle_event(events.Start())
         else:
             yield DropStream(self.stream_id)
 
@@ -749,8 +749,8 @@ class HttpStream(layer.Layer):
 
         if 200 <= self.flow.response.status_code < 300:
             self.child_layer = self.child_layer or layer.NextLayer(self.context)
-            yield from self.child_layer.handle_event(events.Start())
             self._handle_event = self.passthrough
+            yield from self.child_layer.handle_event(events.Start())
             yield SendHttp(
                 ResponseHeaders(self.stream_id, self.flow.response, True),
                 self.context.client,

